### ISSUE TITLE
Tasks with completed signature but missing output files are still ignored.

### ISSUE THREAD

**Issue by BoPeng** [vatlab/sos#1493]
I think the problem here is that `task` does not really check if the output file exists since the tasks could have been executed from an external site. If `output` has been defined and a signature exists, we should at least check the output and print a warning if the output does not exist.

This is not true in most cases and we will need to find a case when signature checking fails.

**Comment by pgcudahy** [comment]
Hello Bo, thanks again for your awesome work on SoS. I think I've run into this problem on my university cluster since one of the filesystems automatically deletes files after 60 days. SoS doesn't always pick up that output files have been deleted so I have to force run jobs if I haven't done them in a while. I think the signatures are stored in the `.sos` directory in my home folder, which is not subject to the periodic deletions.

**Comment by BoPeng** [comment]
SoS automatically removes some old files (mostly tasks) and I cannot recall if these will affect signatures. Let me check

**Comment by BoPeng** [comment]
Could you describe the problem in more detail? There is a known problem with task signatures. That is to say, when you execute a workflow with tasks, the tasks signatures are not checked. This is the case even when  if the steps are executed due to missing `_output`.

The reason for this arrangement is that tasks are external and could be executed on remote hosts, so technically their `_output` do not have to reside on local storage. Only when the tasks are really re-executed do their signatures are checked against their running environment.

I understand that this behavior is not optimal. I can see how I can improve it if you are indeed running such a problem.

**Comment by BoPeng** [comment]
### Test case 1: workflow step level signature

```
[10]
output: 'a.txt'
task:
sh: expand=True
    echo "haha " > {_output}
```

Running
```
sos run test -q test.sos
rm a.txt
sos run test -q test.sos
```
works since `a.txt` does not exist.

### test case 2: task level signature (ignored)

```
[10]
task:

output_file = 'a.txt'
sh: expand=True
    echo "haha " > {output_file}
```

```
(ictr) vatlab/sos - (master) sos run test -q localhost
INFO: Running 10:
INFO: Workflow default (ID=w2c64b5340e5409db) is executed successfully with 1 completed step and 1 completed task.
(ictr) vatlab/sos - (master) rm a.txt
(ictr) vatlab/sos - (master) sos run test -q localhost
INFO: Running 10:
INFO: t9763d971b774326b re-execute completed
INFO: Workflow default (ID=w2c64b5340e5409db) is executed successfully with 1 completed step and 1 completed task.
(ictr) vatlab/sos - (master) ls a.txt
ls: a.txt: No such file or directory
```

The problem here is that the task is said to be `re-executed` and `completed`, but is not executed since

1. It has a signature in task file
2. it does not check the existence of files.

**Comment by BoPeng** [comment]
@pgcudahy Please try the new sos 0.23.3 release and let me know if you continue to notice problems.

**Comment by pgcudahy** [comment]
Hello Bo, I've tried the new release and there still seem to be some issues. I have a directory with 510 files. If I run
```
%run test -q yale_hpc_task_spooler \
-r yale_hpc_task_spooler mem="4GB" cores=2 walltime="04:00:00" nodes=1

[global]
import glob
import os

fastq_files = sorted(glob.glob("/home/pgc29/scratch60/malawi_genomic_epi/dataraw/*fastq.gz"))

[test]
input: fastq_files, group_by=1
output: f'/home/pgc29/scratch60/malawi_genomic_epi/data/test/{_input[0]:bnn}.test'
task: walltime='00:01:00', mem="1GB", cores=1, nodes=1,
    workdir='/home/pgc29/scratch60/malawi_genomic_epi/data/test'

run: expand = True
    mkdir -p {_output:d}
    touch {_output}
```
and then randomly delete one output with `rm TB-852859-204_S133_R2_001.test` and rerun, it works with 
```
$ sos status w259bdbef0b255167 -v4 | tail -n 20
INFO: test (index=497) is ignored due to saved signature
INFO: test (index=498) is ignored due to saved signature
INFO: test (index=499) is ignored due to saved signature
INFO: test (index=500) is ignored due to saved signature
INFO: test (index=501) is ignored due to saved signature
INFO: test (index=502) is ignored due to saved signature
INFO: test (index=503) is ignored due to saved signature
INFO: test (index=504) is ignored due to saved signature
INFO: test (index=505) is ignored due to saved signature
INFO: test (index=506) is ignored due to saved signature
INFO: test (index=507) is ignored due to saved signature
INFO: test (index=508) is ignored due to saved signature
INFO: t420178ce07ef39e7 re-execute completed
INFO: t420178ce07ef39e7 sent path(/home/pgc29/scratch60/malawi_g... to yale_hpc_task_spooler
INFO: t420178ce07ef39e7 submitted to yale_hpc_task_spooler with job id 26993307
INFO: Waiting for the completion of 1 task.
INFO: t420178ce07ef39e7 received /home/pgc29/scratch60/malawi_genomic_epi/data/test/TB-852859...204_S133_R2_001.test from yale_hpc_task_spooler
INFO: test output:   /home/pgc29/scratch60/malawi_genomic_epi/data/test/TB-848753-204_S149_R1_001.test /home/pgc29/scratch60/malawi_genomic_epi/data/test/TB-848753-204_S149_R2_001.test... (510 items in 510 groups)
INFO: Workflow test (ID=w259bdbef0b255167) is executed successfully with 0.0 completed step, 1 completed substep, 1.0 ignored step, 509 ignored substeps and 510 completed tasks.
```

However, with a "trunk" task like 
```
%run test2 -q yale_hpc_task_spooler \
-r yale_hpc_task_spooler mem="4GB" cores=2 walltime="04:00:00" nodes=1

[global]
import glob
import os

fastq_files = sorted(glob.glob("/home/pgc29/scratch60/malawi_genomic_epi/dataraw/*fastq.gz"))

[test2]
input: fastq_files, group_by=1
output: f'/home/pgc29/scratch60/malawi_genomic_epi/data/test/{_input[0]:bnn}_2.test'
task: walltime='00:01:00', mem="10GB", trunk_size=100, trunk_workers=10,
    workdir='/home/pgc29/scratch60/malawi_genomic_epi/data/test'

run: expand = True
    mkdir -p {_output:d}
    touch {_output}
```
and again delete a random output with `rm TB-852859-204_S133_R2_001_2.test` it fails with 
```
$ sos status wa3833f4dbd71429e -v4 | tail
INFO: test2 (index=501) is ignored due to saved signature
INFO: test2 (index=502) is ignored due to saved signature
INFO: test2 (index=503) is ignored due to saved signature
INFO: test2 (index=504) is ignored due to saved signature
INFO: test2 (index=505) is ignored due to saved signature
INFO: test2 (index=506) is ignored due to saved signature
INFO: test2 (index=508) is ignored due to saved signature
INFO: test2 (index=509) is ignored due to saved signature
ERROR: [test2]: [test2]: too many values to unpack (expected 3)
```

**Comment by BoPeng** [comment]
I can reproduce the problem with the following example

```
[get_input: provides='{base}.txt']
sh: expand=True
  echo {_output} > {_output}

[test]
input: [f'a_{i}.txt' for i in range(10)], group_by=1
output: _input.with_suffix('.bak')
task: trunk_size=5

run: expand = True
    cp {_input} {_output}
```

after running

```
sos run test -q localhost
rm a_2.bak
sos run test -q localhost
```

**Comment by BoPeng** [comment]
6ce8fb7e3a271d986afc1f88973af09a3d5b27c1

Will push when the tests pass.

**Comment by BoPeng** [comment]
Fixed again and released sos 0.23.4.
